src/Compilers/Core/Portable/Syntax/GreenNode.cs

```diff
@@ -904,34 +904,61 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
+         * In those cases List<TFrom> performs much better.
+         * In other cases, the type is unknown / is IEnumerable<T>, where we try to find the best match.
+         * There is another overload for IROList, since most collections already implement this, so checking for it will
+         * perform better then copying to a List<T>, though not as good as List<T> directly.
+         */
+        public static GreenNode? CreateList<TFrom>(IEnumerable<TFrom>? enumerable, Func<TFrom, GreenNode> select)
+            => enumerable switch
+            {
+                null => null,
+                List<TFrom> l => CreateList(l, select),
+                IReadOnlyList<TFrom> l => CreateList(l, select),
+                _ => CreateList(enumerable.ToList(), select)
+            };
+
+        public static GreenNode? CreateList<TFrom>(List<TFrom> list, Func<TFrom, GreenNode> select)
         {
-            if (nodes == null)
+            switch (list.Count)
             {
-                return null;
+                case 0:
+                    return null;
+                case 1:
+                    return select(list[0]);
+                case 2:
+                    return SyntaxList.List(select(list[0]), select(list[1]));
+                case 3:
+                    return SyntaxList.List(select(list[0]), select(list[1]), select(list[2]));
+                default: {
+                    var array = new ArrayElement<GreenNode>[list.Count];
+                    for (int i = 0; i < array.Length; i++)
+                        array[i].Value = select(list[i]);
+                    return SyntaxList.List(array);
+                }
             }
+        }
 
-            var list = nodes.ToArray();
-
-            switch (list.Length)
+        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom> list, Func<TFrom, GreenNode> select)
+        {
+            switch (list.Count)
             {
                 case 0:
                     return null;
                 case 1:
-                    if (alwaysCreateListNode)
-                    {
-                        goto default;
-                    }
-                    else
-                    {
-                        return list[0];
-                    }
+                    return select(list[0]);
                 case 2:
-                    return SyntaxList.List(list[0], list[1]);
+                    return SyntaxList.List(select(list[0]), select(list[1]));
                 case 3:
-                    return SyntaxList.List(list[0], list[1], list[2]);
-                default:
-                    return SyntaxList.List(list);
+                    return SyntaxList.List(select(list[0]), select(list[1]), select(list[2]));
+                default: {
+                    var array = new ArrayElement<GreenNode>[list.Count];
+                    for (int i = 0; i < array.Length; i++)
+                        array[i].Value = select(list[i]);
```

Comment:
📝 Even though `list` was already declared as having non-null `GreenNode` items, the new behavior here is slightly different for the case where a null value appeared. I'm not sure if that's a problem or not.